File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/annotation/RCTMGLMarkerView.java
package com.mapbox.rctmgl.components.annotation;

import android.content.Context;
import android.graphics.PointF;
import android.view.View;

import androidx.annotation.NonNull;

import com.mapbox.bindgen.Expected;
import com.mapbox.bindgen.None;
import com.mapbox.geojson.Geometry;
import com.mapbox.geojson.Point;
import com.mapbox.maps.MapboxMap;
import com.mapbox.maps.ViewAnnotationOptions;
import com.mapbox.maps.viewannotation.ViewAnnotationOptionsKtxKt;
import com.mapbox.rctmgl.components.AbstractMapFeature;
import com.mapbox.rctmgl.components.mapview.OnMapReadyCallback;
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;
import com.mapbox.rctmgl.utils.GeoJSONUtils;

public class RCTMGLMarkerView extends AbstractMapFeature implements /*MarkerView.OnPositionUpdateListener, */View.OnLayoutChangeListener {
    private RCTMGLMarkerViewManager mManager;
    private RCTMGLMapView mMapView;

    private View mChildView;

    private Point mCoordinate;
    private Float[] mAnchor;


    public RCTMGLMarkerView(Context context, RCTMGLMarkerViewManager manager) {
        super(context);
        mManager = manager;
    }

    @Override
    public void addView(View childView, int childPosition) {
        mChildView = childView;
    }

    public void setCoordinate(Point point) {
        mCoordinate = point;

        if (mChildView != null) {
            ViewAnnotationOptions options = new ViewAnnotationOptions.Builder().
                    geometry(mCoordinate).
                    build();

            mMapView.getViewAnnotationManager().updateViewAnnotation(mChildView, options);
        }
    }

    public void setAnchor(float x, float y) {
        mAnchor = new Float[]{x, y};
        refresh();
    }

    public void refresh() {
        // this will cause position to be recalculated
        if (mChildView != null) {
            int width = mChildView.getWidth();
            int height = mChildView.getHeight();
            ViewAnnotationOptions options = new ViewAnnotationOptions.Builder().
                    geometry(mCoordinate).
                    width(width).
                    height(height).
                    offsetX((int)((mAnchor[0] - 0.5) * width)).
                    offsetY((int)((mAnchor[1] - 0.5) * height)).
                    build();

            mMapView.getViewAnnotationManager().updateViewAnnotation(mChildView, options);
        }
    }

    @Override
    public void addToMap(RCTMGLMapView mapView) {
        mMapView = mapView;

        final RCTMGLMarkerView rctmglMarkerView = this;

        mMapView.getMapAsync(
            new OnMapReadyCallback() {
                @Override
                public void onMapReady(@NonNull MapboxMap mapboxMap) {
                    if (mChildView != null) {
                        GeoJSONUtils.toLatLng(mCoordinate);

                        int width = mChildView.getWidth();
                        int height = mChildView.getHeight();

                        ViewAnnotationOptions options = new ViewAnnotationOptions.Builder().
                                geometry(mCoordinate).
                                width(width).
                                height(height).
                                offsetX((int)((mAnchor[0] - 0.5) * width)).
                                offsetY((int)((mAnchor[1] - 0.5) * height)).
                                build();

                        mChildView.addOnLayoutChangeListener(rctmglMarkerView);
                        if (mChildView.getLayoutParams() == null && !mChildView.isAttachedToWindow()) {
                            mMapView.offscreenAnnotationViewContainer().addView(mChildView);
                            mMapView.offscreenAnnotationViewContainer().removeView(mChildView);
                        }
                        mMapView.getViewAnnotationManager().addViewAnnotation(mChildView, options);
                    }
                }
            }
        );
    }

    @Override
    public void removeFromMap(RCTMGLMapView mapView) {
        if (mChildView != null) {
            mMapView.getViewAnnotationManager().removeViewAnnotation(mChildView);
            mChildView.removeOnLayoutChangeListener(this);
        }
    }

    @Override
    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop,
                               int oldRight, int oldBottom) {
        if (left != oldLeft || right != oldRight || top != oldTop || bottom != oldBottom) {
            refresh();
        }
    }
}

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/annotation/RCTMGLMarkerView.kt
package com.mapbox.rctmgl.components.annotation

import android.content.Context
import android.util.Log
import android.view.View
import com.mapbox.geojson.Point
import com.mapbox.maps.ViewAnnotationAnchor
import com.mapbox.maps.ViewAnnotationOptions
import com.mapbox.maps.viewannotation.viewAnnotationOptions
import com.mapbox.rctmgl.components.AbstractMapFeature
import com.mapbox.rctmgl.components.mapview.RCTMGLMapView
import java.util.Vector

private data class Vec2(val dx: Double, val dy: Double)

class RCTMGLMarkerView(context: Context?, private val mManager: RCTMGLMarkerViewManager):
    AbstractMapFeature(context),
    View.OnLayoutChangeListener
{
    // region Instance variables

    private var mMapView: RCTMGLMapView? = null
    private var mView: View? = null
    private var didAddToMap = false

    private var mCoordinate: Point? = null
    private var mAnchor: Vec2 = Vec2(0.5, 0.5)
    private var mAllowOverlap = false
    private var mIsSelected = false

    fun setCoordinate(point: Point?) {
        mCoordinate = point
        update()
    }

    fun setAnchor(x: Float, y: Float) {
        mAnchor = Vec2(x.toDouble(), y.toDouble())
        update()
    }

    fun setAllowOverlap(allowOverlap: Boolean) {
        mAllowOverlap = allowOverlap
        update()
    }

    fun setIsSelected(isSelected: Boolean) {
        mIsSelected = isSelected
        update()
    }

    // endregion

    // region View methods

    override fun addView(childView: View, childPosition: Int) {
        mView = childView
        // Note: Do not call this method on `super`. The view is added manually.
    }

    override fun onLayoutChange(
        v: View,
        left: Int, top: Int, right: Int, bottom: Int,
        oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int
    ) {
        addOrUpdate()
    }

    // endregion

    // region RCTMGLMapComponent methods

    override fun addToMap(mapView: RCTMGLMapView) {
        mMapView = mapView
        add()
    }

    override fun removeFromMap(mapView: RCTMGLMapView) {
        remove()
    }

    // endregion

    // region Create, update, and remove methods

    private fun addOrUpdate() {
        if (didAddToMap) {
            update()
        } else {
            add()
        }
    }

    private fun add() {
        if (didAddToMap) {
            return
        }

        if (mView == null || mCoordinate == null) {
            return
        }
        val view = mView!!

        view.addOnLayoutChangeListener(this)

        if (view.layoutParams == null && !view.isAttachedToWindow) {
            mMapView?.offscreenAnnotationViewContainer?.addView(view)
            mMapView?.offscreenAnnotationViewContainer?.removeView(view)
        }

        val options = getOptions()

        val annotation = mMapView?.viewAnnotationManager?.addViewAnnotation(
            view,
            options
        )
        didAddToMap = true
    }

    fun update() {
        if (!didAddToMap) {
            return
        }

        if (mView == null || mCoordinate == null) {
            return
        }
        val view = mView!!

        val options = getOptions()

        val annotation = mMapView?.viewAnnotationManager?.updateViewAnnotation(
            view,
            options
        )
    }

    private fun remove() {
        this.removeOnLayoutChangeListener(this)

        if (mView == null) {
            return
        }

        val removed = mMapView?.viewAnnotationManager?.removeViewAnnotation(mView!!)
        if (removed == false) {
            Log.d("[MarkerView]", "Unable to remove view")
        }
    }

    // endregion

    // region Helper functions

    private fun getOptions(): ViewAnnotationOptions {
        val view = mView!!
        val width = view.width
        val height = view.height
        val coordinate = mCoordinate

        val offset = getOffset()

        val options = viewAnnotationOptions {
            geometry(coordinate)
            width(width)
            height(height)
            allowOverlap(mAllowOverlap)
            offsetX(offset.dx.toInt())
            offsetY(offset.dy.toInt())
        }
        return options
    }

    private fun getOffset(): Vec2 {
        if (mView == null) {
            return Vec2(0.0, 0.0)
        }
        val view = mView!!

        val width = view.width
        val height = view.height

        // Create a modified offset:
        // - Normalize from [(0, 0), (1, 1)] to [(-1, -1), (1, 1)].
        // - Scale to the view size.
        // - Invert `y` so that higher values are lower on the screen.
        val offsetX = (mAnchor.dx * 2 - 1) * (width / 2)
        val offsetY = (mAnchor.dy * 2 - 1) * (height / 2) * -1

        return Vec2(offsetX, offsetY)
    }

    // endregion
}